### PR TITLE
[SPARK-13435] [MLlib] Add Weighted Cohen's kappa to MulticlassMetrics

### PR DESCRIPTION
**Pull request by zhengruifeng**
JIRA: https://issues.apache.org/jira/browse/SPARK-13435
## What changes were proposed in this pull request?

Add the missing Weighted Cohen's kappa to MulticlassMetrics.
## How was the this patch tested?

unit tests and manual tests were done. 
The calculation correctness was tested on several small data, and compared to two python's package: sklearn and ml_metrics.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 1]
```diff
@@ -25,18 +25,19 @@ import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.DataFrame
 
 /**
- * ::Experimental::
- * Evaluator for multiclass classification.
- *
- * @param predictionAndLabels an RDD of (prediction, label) pairs.
- */
+  * ::Experimental::
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Nit: undo these space changes

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 2]
```diff
@@ -129,86 +135,199 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
...
+    * Returns unweighted Cohen's Kappa
+    * Cohen's kappa coefficient is a statistic which measures inter-rater
+    * agreement for qualitative (categorical) items. It is generally thought
+    * to be a more robust measure than simple percent agreement calculation,
+    * since kappa takes into account the agreement occurring by chance.
+    * The kappa score is a number between -1 and 1. Scores above 0.8 are
+    * generally considered good agreement; zero or lower means no agreement
+    * (practically random labels).
+    */
+  @Since("1.6.0")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: This can't be right since 1.6.0 is out. 2.0.0.
Your scaladoc indentation is off and you need to not automatically alter the rest of it. Most of your PR is whitespace change

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 3]
```diff
@@ -129,86 +135,199 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
...
+      case "default" =>
+        (i: Int, j: Int) => {
+          if (i == j) {
+            0.0
+          } else {
+            1.0
+          }
+        }
+      case "linear" =>
+        (i: Int, j: Int) => Math.abs(i - j).toDouble
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Nit: `math.abs`

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 4]
```diff
@@ -129,86 +135,199 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
...
+      case "quadratic" =>
+        (i: Int, j: Int) => (i - j).toDouble * (i - j)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Also tiny nit, but this needlessly computes the difference twice

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 5]
```diff
@@ -129,86 +135,199 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
...
+
+
+  /**
+    * Returns Cohen's Kappa with user-defined weight matrix
+    *
+    * @param weights the weight matrix, must be of the same shape with Confusion Matrix.
+    *                Note: Each Element in it must be no less than zero.
+    */
+  @Since("1.6.0")
+  def kappa(weights: Matrix): Double = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: I don't think this is that practical to expose; the matrix would in most cases need to be far too large. At least, this doesn't seem like the right thing to expose in a public API

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala [hunk 6]
```diff
@@ -51,6 +51,9 @@ class MulticlassMetricsSuite extends SparkFunSuite with MLlibTestSparkContext {
     val f2measure0 = (1 + 2 * 2) * precision0 * recall0 / (2 * 2 * precision0 + recall0)
     val f2measure1 = (1 + 2 * 2) * precision1 * recall1 / (2 * 2 * precision1 + recall1)
     val f2measure2 = (1 + 2 * 2) * precision2 * recall2 / (2 * 2 * precision2 + recall2)
+    val unweighted_kappa = 0.47058823529411764
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala]: Out of curiosity did you obtain these from something like R? that's always great as a double check if possible

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: Sorry, this is caused by my IDE's reformating. I will undo those.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 4 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: ok, I will fix it

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 5 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: I trend to keep this api, for the input matrix is small, whose size is nClass * nClass.
And in the ﻿relative statistical documents (such as http://www.real-statistics.com/reliability/weighted-cohens-kappa/), there exists the weighted matrix. Keeping it may make kappa score more comprehensive.

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala (hunk 6 above)
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala]: The three number are obtained via sklearn and ml_metrics.

In [4]: from sklearn.metrics import cohen_kappa_score

In [5]: from ml_metrics import quadratic_weighted_kappa, linear_weighted_kappa, kappa

In [6]: preds = [0, 0, 0, 1, 1, 1, 1, 2, 2]

In [7]: labels = [0, 1, 0, 0, 1, 1, 1, 2, 0]

In [8]: cohen_kappa_score(preds, labels)
Out[8]: 0.47058823529411781

In [9]: quadratic_weighted_kappa(preds, labels)
Out[9]: 0.3571428571428571

In [10]: linear_weighted_kappa(preds, labels)
Out[10]: 0.4193548387096774

In [11]: kappa(preds, labels)
Out[11]: 0.47058823529411764

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 3 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: ok, I have fix it

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 2 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: ok, I have fix it

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 1 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: It take me half one day to refomart the code... 
I use IntelliJ IDEA CE as main IDE, and where can I find the correct IDEA Code Style config for spark?

File: mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala (hunk 6 above)
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/mllib/evaluation/MulticlassMetricsSuite.scala]: Sounds good to me

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala [hunk 7]
```diff
@@ -211,4 +211,119 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
...
+  }
+
+
+  /**
+   * Returns Cohen's Kappa with user-defined weight matrix
+   * @param weights the weight matrix, must be of the same shape with Confusion Matrix.
+   *                Note: Each Element in it must be no less than zero.
+   */
+  @Since("2.0.0")
+  def kappa(weights: Matrix): Double = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: I still don't think we should expose this. I can be private.

**Comment by srowen** [comment]
Jenkins, test this please

**Comment by srowen** [comment]
Any thoughts @mengxr ?

As for code formatting, I don't think IJ will change lines you don't touch. However you can edit the style settings in the editor preferences, and just make it write the scaladoc without the extra level of indent.

I wouldn't worry too much about the style -- just follow what you see around your change and undo changes that aren't related

**Comment by zhengruifeng** [comment]
@srowen yes, the style problem was caused by that the IJ's default 'reformating' style for scala dont match spark's.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala (hunk 7 above)
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala]: ok, this place needs more discussion.
I still think it appropriate to view the weights as a matrix.

**Comment by zhengruifeng** [comment]
cc @jkbradley @yanboliang
